src/controllers/resource-store/index.tsx

```diff
@@ -189,72 +192,85 @@ export const privateActions = {
 
           dispatch(validateLRUCache(resource, key));
 
-          const pendingSlice = {
-            ...prevSlice,
-            data: maxAge === 0 ? null : prevSlice.data,
-            error: maxAge === 0 ? null : prevSlice.error,
-            loading: true,
-            promise: getData(
-              { ...routerStoreContext, isPrefetch: !!prefetch },
-              context
-            ),
-            accessedAt: getAccessedAt(),
-            // prevent resource from being cleaned prematurely and traigger more network requests.
-            ...(options.prefetch
-              ? {
-                  expiresAt: getExpiresAt(PREFETCH_MAX_AGE),
-                }
-              : {}),
-          };
-
-          dispatch(setResourceState(type, key, pendingSlice));
-
-          const response = {
-            ...pendingSlice,
-          };
-
-          try {
-            response.error = null;
-
-            if (timeout) {
-              const timeoutGuard = generateTimeGuard(timeout);
-              const maybeData = await Promise.race([
-                pendingSlice.promise,
-                timeoutGuard.promise,
-              ]);
+          const promiseOrData = getData(
+            {
+              ...routerStoreContext,
+              isPrefetch: !!prefetch,
+            },
+            context
+          );
 
-              if (!timeoutGuard.isPending) {
-                response.data = null;
-                response.error = new TimeoutError(type);
-                response.loading = true;
-                response.promise = null;
-              } else {
-                timeoutGuard.timerId && clearTimeout(timeoutGuard.timerId);
-                response.data = maybeData;
-                response.loading = false;
-              }
-            } else {
-              response.data = await pendingSlice.promise;
-              response.loading = false;
+          let resolvedSlice;
+
+          if (promiseOrData === prevSlice.data) {
+            // same data (by reference) means nothing has changed and we can avoid loading state
+            resolvedSlice = {};
+          } else {
+            // ensure the promise includes any timeout error
+            const timeoutGuard = timeout ? generateTimeGuard(timeout) : null;
+            const promise = timeout
+              ? Promise.race([promiseOrData, timeoutGuard?.promise]).then(
```

Comment:
This `promiseOrData` needs to be `toPromise(promiseOrData)` otherwise `Promise.race()` will ignore it.

Reply:
actually I retested and this seems fine, I will leave as is and ensure it is covered by tests